src/ui/public/timepicker/timepicker.js

```diff
@@ -76,11 +76,19 @@ module.directive('kbnTimepicker', function (quickRanges, timeUnits, refreshInter
       });
 
       $scope.$watch('absolute.from', function (date) {
-        if (_.isDate(date)) $scope.absolute.from = moment(date);
+        // If it is a date, it was set from the timepicker, so we set it to the start of day and transform it into a moment
+        if (_.isDate(date)) {
+          date.setHours(0, 0, 0, 0); // Start of day
```

Comment:
Instead of having to guess where the update came from, what if we used a getterSetter for the ng-model on the datepicker using [ngModelOptions](https://docs.angularjs.org/api/ng/directive/ngModelOptions)? Then this `setHours` logic could live inside the setter.